### ISSUE TITLE
Slick 3.0 withTransaction blocks are required to interact with libraries

### ISSUE THREAD

**Issue by Ciantic** [slick/slick#1197]
There are at least two #1148 and #1093 that looks like a duplicate, but according to discussion at Slack with @cvogt I decided to open this anyway.

Suppose there is a *external library* which has trait e.g.

```
trait CatDao {
  def save(input: Cat): Future[Cat]
}
trait DogDao {
  def save(input: Dog): Future[Dog]
}
```

If I then create implementation using slick for A, and B, they look something like this:

```
class DbCatDao extends CatDao {
  def save(input: Cat): Future[Cat] = {
    db.run(((Table returning Table.map(_.id) into ((u, insertId) => input.copy(id=insertId))) += input))
  }
}

class DbDogDao extends DogDao {
  def save(input: Dog): Future[Dog] = {
    db.run(((Table returning Table.map(_.id) into ((u, insertId) => input.copy(id=insertId))) += input))
  }
}
```

Now somewhere in code you may need to call them both in one transaction, it was possible with `db.withTransaction` before, now it is not possible at all because it was deprecated.

```
db.withTransaction {
  for {
    cat <- dbCatDao.save(...)
    dog <- dbDogDao.save(...) 
  }
}
```

With having only `transactionally` available, there is no good way to fix this, because they don't return `DBIOAction`, and I can't change their return types because they are meant to implement the contract in given in external library.

I understand that `transactionally` is better when one controls the libraries, but it's not suitable when interacting with external libraries always.

**Comment by johanstenberg92** [comment]
+1 I'm in the same boat. I need to have large transactions, I can't really understand how this is done in Slick 3.0.1.

**Comment by i-schuetz** [comment]
What I do in these cases is having public methods that return actions.

**Comment by johanstenberg92** [comment]
Can you give some example code?

**Comment by schmitch** [comment]
That's enterily possible:

    db.run{
     (for {cat <- dbCatDao.save(); dog <- dbDogDao.save() } yield ()).transactionally
   }

Currently if you have some DBIO.from(Future{})  you need to have withPinnedsession oh and dbCatDao.save() just needs everything without the wrapped db.run()

**Comment by i-schuetz** [comment]
```
class DbCatDao extends CatDao {
  def saveAction(input: Cat): DBIO[Cat] = {
    ((Table returning Table.map(_.id) into ((u, insertId) => input.copy(id=insertId))) += input)
  }

  def save(input: Cat): Future[Cat] = {
    db.run(saveAction(input))
  }
}
```

**Comment by schmitch** [comment]
@i-schuetz DBIO.from is only for things that doesn't access the database, Like Create a database entry and also update Elasticsearch
The thing how you did val action is correct. I just wanted to point out that when people do other DBIO's while running a transaction they need a pinned session. But it could be really helpful if you want to keep things in sync, etc.

**Comment by asazernik** [comment]
I've made this both work and be quite easy on the eye by adding a couple of implicit conversions; you can do this all manually, but it'll be ugly.

```scala
object AsyncUtils {

  implicit def tryToFuture[A](t: Try[A]): Future[A] = t match {
    case Success(s) => Future.successful(s)
    case Failure(e) => Future.failed(e)
  }

  implicit def futureToDBIOAction[A](f: Future[A]): DBIOAction[A, NoStream, Effect] =
    DBIO.from(f)

  implicit def tryToDBIOAction[A](t: Try[A]): DBIOAction[A, NoStream, Effect] =
    futureToDBIOAction(t)

}
```

This lets me write code that mixes DBIOActions and Futures and Trys in a flatMap chain or for-comprehension, and the end result will be of the type that is "highest" in the chain Try->Future->DBIOAction. For example, here's some code from my current project:

```scala
    db.run(
      r.userFetch
      .flatMap(userId => supplierFromJson(r.request.body, userId)) // Try, implicitly converted
      .flatMap(s => (Suppliers returning Suppliers.map(_.id)) += s)
      .flatMap { id =>
        val contacts = r.request.body.asInstanceOf[JsObject].value.getOrElse("contacts", JsArray())
        supplierContactsFromJson(contacts, id).map((id, _)) // pass the ID on through; Try, implicitly converted
      }
      .flatMap { case (id, contacts) =>
        // want to abort on failure, return ID on success (even with no contacts)
        (SupplierContacts ++= contacts).andThen(DBIO.successful(id))
      }.transactionally
    )
```

I'm using Trys here, but if you're accessing an asynchronous API a Future will do just as well.

Unfortunately, the compiler doesn't quite do the right implicit conversions if I try to convert this to a much-more-readable for comprehension, which would let me avoid the manual passthrough of variables in this pipeline. You can do that, but only with explicit conversions:

```scala
val massiveOperation = for {
      userId <- r.userFetch
      s <- tryToDBIOAction(supplierFromJson(r.request.body, userId))
      supplierId <- (Suppliers returning Suppliers.map(_.id)) += s
      contactsJson = r.request.body.asInstanceOf[JsObject].value
                     .getOrElse("contacts", JsArray()) // plain computation
      contacts <- tryToDBIOAction(supplierContactsFromJson(contactsJson, supplierId))
      contactsAdded <- SupplierContacts ++= contacts
    } yield supplierId

    db.run(massiveOperation.transactionally)
```

**Comment by i-schuetz** [comment]
@asazernik How does `DBIO.from` work for a transaction? Doesn't `db.run(action)` trigger the database access immediately? If a database operation is executed to update Cat and separate one to update Dog,  even if you convert the futures back to DBIO using `DBIO.from` there can't be a transaction (correct me if I'm wrong).

**Comment by asazernik** [comment]
@i-schuetz Disclaimer: not a Slick developer, I may be very wrong about implementation details.

A better way to think about it is that a DBIOAction like a Future, but depends on a database connection (which `db.run()`) *in addition* to the regular ExecutionContext required by a plain Future. Otherwise, they handle pretty much the same. For example, a Future composed with `flatMap` and `map` and `andThen` - all the same methods DBIOAction implements - only starts the later items in the pipeline when the earlier ones return. That's why these don't take concrete Futures or DBIOActions as arguments, but rather functions that *can be called to produce* the relevant Future/DBIOAction.

In this example (a `flatMap()` chain), db.run() starts the *first* DBIOAction in the chain immediately, and sets up a completion callback that will continue the chain when the `DBIOAction` returns a value from the database (actually, `andThen()` is responsible for doing this - implementation detail). And in an interesting twist, `transactionally` is defined like this (re-factored for clarity, including changing implicit-conversion variable `this`):

```scala
// StartTransaction, Commit, and Rollback are all DBIOActions
def transactionally: DBIOAction = {
  StartTransaction // the real first DBIOAction - executed *immediately* on db.run()
  .andThen(this) // executed only after StartTransaction returns
  .cleanUp(maybeError => if (maybeError.empty) Commit else Rollback) // note - check for error is done *in Scala*
}
```

(The fact that asynchronous client-side code is an equal citizen in all this is also easily observable in the guarantees a DBIOAction *doesn't* make - namely, that chained DBIOActions aren't guaranteed to run on the same DB connection. This implies values being transmitted back to Scala, and then Scala code running asynchronously in response to each step in the pipeline completing. In fact, `transactionally` is just a more extreme version of `withPinnedSession`, which is the only way to force the Scala-side code that runs a chain of DBIOActions to use the same session.)

**Comment by i-schuetz** [comment]
@asazernik 

So far my understanding `DBIO` and `Future` handling is the same, because both are implemented as monads, with the same "output", but they still have an entirely different meaning/usage. When you chain actions, you basically are constructing a database operation which will be executed in one single access to the database. This operation will be executed (asynchronously) when you call `db.run(query)`, which returns a future. You can then compose this future with other futures, use map/flatMap on it, etc. but that's only the future, which represents a result of the operation you already did on the database. The result may not be there yet, but the operation is already initiated and can't be combined anymore with other actions. If you create a new DBIO from a future using `DBIO.from` you will get a new action, but it can't be combined or executed transactionally with the action that originated the future.

As I said this is only my understanding, may be very wrong as I have little experience with Slick. In that case hopefully this conversation server a learning purpose.

**Comment by asazernik** [comment]
They [are not executed in a single DB access](http://slick.typesafe.com/doc/3.0.0/dbio.html#transactions-and-pinned-sessions):

> When executing a DBIOAction which is composed of several smaller actions, Slick acquires sessions from the connection pool and releases them again as needed so that a session is not kept in use unnecessarily while waiting for the result from a non-database computation (e.g. the function passed to flatMap that determines the next Action to run).

**Comment by asazernik** [comment]
To use the terminology of that paragraph: `DBIO.from` creates a new DBIOAction that consists *only* of "non-database computation".

**Comment by i-schuetz** [comment]
I edited my comment at the same time you wrote the answer - yes, it may acquire multiple connections but this is an implementation detail. You can't make any assumptions about this.

My point is, coming back to the original question, that you can't just use an implicit that converts futures to actions in order to execute the Cat and Dog inserts transactionally. This would execute a transaction of 2 actions that represent "non-database computations", which doesn't make sense. It will compile, and the items will be inserted, but not transactionally.

**Comment by asazernik** [comment]
DBIOActions are *not* entirely different in meaning and usage from Futures; they're computations that can be converted into Futures by just telling them which database to operate on (which they might ignore, if they were created with `DBIO.failed`, `DBIO.successful`, or `DBIO.from`). For more insight, take a look at `DatabaseDef.runInContext` (`db.run` is just a thin wrapper on top of this). This is written as a long match statement, but it basically says:

* If this is a DBIOAction that just represents an immediate (synchronous) success/failure, create an immediately-fulfilled Future with that success or failure.
* If this is wrapping a Future (i.e. is from `DBIO.from`), return the wrapped Future.
* If this is a single DBIOAction that *does* require the database, create a new Future for the computation of sending the action to the database and retrieving the results
* **The one you care about**: if this is a DBIOAction created by combining two DBIOActions with andThen or flatMap, convert the first one into the future, and apply the corresponding Future method to the rest of the DBIOAction (which may be another created with a combiner).

So if you look at that previous expansion of `transactionally`, you'll see that it basically works out to

```scala
executeInDB("START TRANSACTION;") // Future
.andThen({
  // execute arbitrary code, with reference to the same database config
})
.cleanUp(error => if (error.empty) executeInDB("COMMIT") else executeInDB("ROLLBACK"))
```

On to your point about transactions: a transaction can absolutely include (i.e. protect the concurrency of) non-database computations. Imagine you wanted to do this:

```sql
START TRANSACTION;

SELECT num FROM bogus_table;
// in client: calculate (num + 1), call it newNum
INSERT $newNum INTO other_bogus_table;
COMMIT
```

This is the "normal" way that you would do things like call out to an external HTTP service in the middle of a transaction; and it will work *just as correctly* as doing the stuff in the database, as long as you somehow enforce ordering on the non-database actions (usually this is done by waiting for the database to return success for the 'START TRANSACTION' statement). In fact, this is how a transactional SQL database can be in general to implement mutexes and locking for the rest of a general-purpose program.

Going back to the Cat/Dog example, the code would look like:

```scala
trait CatDAO {
  saveAction(input: Cat): DBIOAction[Cat, _, _]
}

trait DogDAO {
  saveAction(input: Dog): DBIOAction[Dog, _, _]
}

val catDAO: CatDAO = ???
val dogDAO: DogDAO = ???

db.run(
  catDAO.save(c).andThen(dogDAO.save(d)).transactionally
);
// also expressible as (catDAO.save(c) >> dogDAO.save(d)).transactionally
```

The interesting things to note are that:

1. `transactionally` needs to be applied *before* converting a DBIOAction into a Future with `db.run()`. This is because otherwise the database can do all kinds of re-ordering and operating on temporary views that will destroy ordering guarantees.
2. `transactionally` needs to be applied to the `combined` DBIOAction, not to the individual ones. Applying to both [is a no-op](http://slick.typesafe.com/doc/3.0.0/api/index.html#slick.driver.JdbcActionComponent$JdbcActionExtensionMethods@transactionally:DBIOAction[R,S,EwithTransactional]): "If multiple transactionally combinators are nested, only the outermost one will be backed by an actual database transaction."

**Comment by i-schuetz** [comment]
Well, your explanation matches for the most part what I'm saying and your code comes down to the same what I suggested in my first comment.

In your initial comment you did *not* write that the issue opener has to return actions instead of futures. Instead you suggested to use implicits to convert between actions and futures and basically mix everything together:

>This lets me write code that mixes DBIOActions and Futures and Trys in a flatMap chain or for-comprehension, and the end result will be of the type that is "highest" in the chain Try->Future->DBIOAction

Which would, as yourself describe yourself not work for a transaction. It only works if the methods return actions and then the implicit is also not necessary.

Also, saying that DBIOActions are not entirely different in meaning and usage from Futures because running a DBIOAction can lead to future / is converted to a future is like saying the input of a function is not different from the output because one leads to the other, don't really see a meaning in that. They are  related, in the context of the function/library but still different things.

Thanks tough for the additional details, specially about transactions being able to include non-database computations. Didn't know that.

**Comment by asazernik** [comment]
Ah. The confusion is, I was trying to address the issue of using non-Slick libraries in a transaction context. The implicit conversions should only be used (and the conversions whether implicit or not are general are only well-defined) in the *upward* direction - that is, convert Trys and Futures (that are *not* originally DBIOActions) into DBIOActions, and then mix all of *those* together. i.e. Future is a superset of Try, and DBIOAction is a superset of Future. Not super relevant to the Dog/Cat example, but it seemed like that was where the conversation had headed.

But yeah, bringing that back to the concrete example, the general principle is to convert upwards and never downwards for as long as possible, and then running `db.run()` once on the accumulated DBIOAction.

**Comment by i-schuetz** [comment]
IIRC at the beginning the formulation of the question was a bit broader, not restricted to only external libraries. 

And anyway, how do those solutions help if it was an external library? If the actions are not exposed, there's no way to execute them in a transaction. Wouldn't that be a problem of the library? If the authors decide to define the api such that these transactions are not possible there's maybe a reason. I would contact the authors and ask either for an entry point that does the transaction or to expose the actions...

**Comment by i-schuetz** [comment]
The question is about implementing a trait of an external library. You can create a new trait that inherits from that trait, which adds the methods that return `DBIOAction`, which you can then use in your own code, to do the transactions. Then implement the library's methods such that they also use the methods that return `DBIOAction`. Which is basically the code I posted (I put the methods in the classes, but they can be in traits as well obviously). Not sure if I'm missing something, but for me it looks like this would solve the issue.

**Comment by asazernik** [comment]
If an external library defined its API like this, then it is broken and needs to be fixed.

In more detail: if a library is doing something that operates on a database, and the user wants those operations to be in the same transaction as their own operations, then some arrangement needs to be made so that the library and user code are using the same connection. Otherwise all transactional consistency guarantees are out the window.

So there needs to be some agreement between user and creator of the library. In Slick 2.x, you did this by making sure to have the library functions take an `implicit session: Session` parameter. in the Slick 3.x case, that agreement is that libraries give you DBIOActions. (Unless the library is not touching the database, in which case you *can* just use the `Future[Result]`. with `DBIO.from`.) If an external library does not do this, then it is broken and will trash your data on massively concurrent access, just like if a Slick 2.x library tried to get a new session instead of taking the `implicit session` parameter from its caller.

**Comment by i-schuetz** [comment]
I would not be so quick to say it's broken. Probably it is, but we don't know the concrete example at hand. It totally depends what this library wants to support / expose, and also if this is a Slick (2.x/3.x) or non Slick library. As I said, I would contact the authors and talk about it.

One way or the other the trait can be extended as I described, as long as this can be made available for the code that needs to execute the transactions. If this functionality has to be in the library's contract, or not, as said, depends on the concrete use case.

**Comment by cvogt** [comment]
@i-schuetz are you suggesting

```scala
// unmodifiable traits from library
trait CatDao {
  def save(input: Cat): Future[Cat]
}
trait DogDao {
  def save(input: Dog): Future[Dog]
}

// implementation
class DbCatDao extends CatDao {
  def save(input: Cat): Future[Cat] = db.run(saveDBIO)
  def saveDBIO(input: Cat): DBIO[Cat] = {
    (Table returning Table.map(_.id) into ((u, insertId) => input.copy(id=insertId))) += input
  }
}

class DbDogDao extends DogDao {
  def save(input: Dog): Future[Dog] = db.run(saveDBIO)
  def saveDBIO(input: Dog): DBIO[Dog] = {
    (Table returning Table.map(_.id) into ((u, insertId) => input.copy(id=insertId))) += input
  }
}

// transaction
db.withTransaction {
  for {
    cat <- dbCatDao.saveDBIO(...)
    dog <- dbDogDao.saveDBIO(...) 
  }
}
```

**Comment by i-schuetz** [comment]
@cvogt yes

**Comment by cvogt** [comment]
@i-schuetz nothing wrong with it, sounds like a good way to go in such a case. Not sure you can't construct an example where this breaks down, but I don't see a super obvious one right now.

**Comment by szeiger** [comment]
Marking this issue for use in the planned documentation improvements in 3.2. Just a quick comment about composition at this point: The distinction between methods that can be composed in a transactional way and those that cannot has not changed with the new API. Any method that did not take a `Session` was not composable before. When you have an old method of the form

```def f(...)(implicit session: Session): T```

it should be changed to

```def f(...): DBIO[T]```

to preserve composition semantics.

**Comment by pbatko** [comment]
@szeiger have you considered providing overloaded `db.run` accepting additional `session` parameter so that transaction could be preserved across multiple database calls using old style methods:

    def f(...)(implicit session: Session): T
This would ease up migrating to the new API. Without this I'm forced to change entire application (thousands of classes) to the new style methods in the instant of upgrading to Slick 3.

Although Slick 3 preserves `db.withTransaction` API (only deprecating it), it is useless because transaction semantics across multiple old style db calls is no longer preserved.

**Comment by pbatko** [comment]
@szeiger @cvogt Is there a way (even deprecated one) to keep transactional semantics without being forced to change whole application into `map`s and `flatMap`s on `DBIOAction`?

**Comment by cvogt** [comment]
@szeiger I heard these troubles of the migration being a bit rough for some projects multiple times. @szeiger how hard would it be to support or write a gist to allow people to pass their own sessions/connection through a series of actions, i.e. offer a db.runWithSession(dbio)(session)?

**Comment by pbatko** [comment]
@cvogt @szeiger I came up with a implementation of `db.runWithSession` with Slick 3.0.3.
You can see the code here: https://github.com/pbatko/slick/commits/db-run-with-session
It seems to work, though I'm not sure if I covered all use cases.

Here is an example usage: https://github.com/pbatko/slick3-migration-playground

    def invokeAction[R, S <: slick.dbio.NoStream, E <: slick.dbio.Effect](action: DBIOAction[R, S, E])
                                                                       (implicit session: Session): R = {
        val db = session.database
        Await.result(db.runWithSession(action, session), queryTimeout)
    }
I had some trouble with importing the session. The import I was usually using was getting me a type projection whereas `runWithSession` expects a path-dependent type. I solved it by shuffling the imports like this:

    //Following import causes an error:
    /*
    [error]  found   : slick.driver.PostgresDriver.api.Session
    [error]     (which expands to)  slick.jdbc.JdbcBackend#SessionDef
    [error]  required: _13.SessionDef where val _13: slick.jdbc.JdbcBackend
    [error]     Await.result(db.runWithSession(action, session), queryTimeout)
     */
    //import slick.driver.PostgresDriver.api._
    
    //This works
    import slick.driver.PostgresDriver.api.{Session => _, _}
    import slick.driver.PostgresDriver.backend.Session

Does this feature have chances of getting merged into Slick?

**Comment by cvogt** [comment]
As this seems to be a common migration problem, I support the integration of this feature. Probably it should be documented that this is not the recommended way to use Slick. @szeiger can better judge the correctness of the commit. @pbatko can you make a PR out of it please, so we can discuss the code there?

**Comment by pbatko** [comment]
@cvogt I made PR https://github.com/slick/slick/pull/1315

**Comment by mikebridge** [comment]
Do these changes for v3 mean that it is no longer possible to write the equivalent of Setup/TearDown code in an integration test so that I can open & roll back the DB modifications for each Test?  From what I gather, the only way to wrap anything in transactions is to compose DBIOActions, then use .transactionally (and .withPinnedSession) to execute them.  I don't see a way of using an open TX in a context....

I was trying to write a ForEach trait in specs2 that would roll back each spec, but it's not clear to me whether it can still be done without wrapping each part in a DBIO, which isn't what I want.

I've just started to use slick, but I got stuck on this....

**Comment by chadjaros** [comment]
Like @mikebridge comments, I'm currently trying to set up a project using Slick 3.1, MySQL, Flyway, and H2 for testing. Using H2 for database access, I can simply tear down the database between test methods.

The combination of Flyway, MySQL/H2, and Slick seems to be broken on a pretty fundamental level, and one of my fallback solutions is to use something other than H2 as the testing database. Any database which I can't tear down for every test is in need of some way to manage auto-rollback transactions for each test case, and I can't imagine how to accomplish that with the current transaction management paradigm.

**Comment by mikebridge** [comment]
@chadjaros You can see my best attempt to synthesize this and other threads to roll back transactions cleanly [on stack overflow](http://stackoverflow.com/questions/34905455/how-can-i-roll-back-an-integration-test-with-slick-3-specs2).  It's ugly, but I believe you cannot roll back transactions without wrapping all DB and non-DB actions that happen during a transaction inside DBIO monads, binding them together, then throwing and discarding an exception in a final monad and then going back and grabbing the original output.

This seemed to work anecdotally, but if they run asynchronously I'm really not sure what happens if the exception gets thrown before the real queries happen. The sheer complexity of trying to write something that should be trivial like an integration test made me realize that that the philosophy behind slick doesn't line up well with needs of the projects I am working on, and I had to move to something that was a little more straightforward.

**Comment by chadjaros** [comment]
@mikebridge Yeah, I'm not sure that it's a dealbreaker for my projects yet. I'm brute forcing it for the moment by re-initializing the schema for each test. It's working for now, and I'm cautiously optimistic that I won't have to fall back to something like scalikejdbc.

**Comment by mikebridge** [comment]
@chadjaros A few people on gitter were dropping and recreating everything with each test but there were no official recommendations.

**Comment by schmitch** [comment]
@mikebridge sometimes you need that, because on postgresql you can't start a new transaction if your application already has a transaction in place so rolling back wouldn't work if a commit happend already.

**Comment by mikebridge** [comment]
@schmitch Hmm, that may be another reason for me to stick with a more jdbc-native abstraction like jooq so that I can access postgresql-specific features like savepoints if I need them.  

There are hints that it is possible to drop down to JDBC in slick but I couldn't figure it out.

**Comment by schmitch** [comment]
@mikebridge savepoints will not help you if you have a running transaction.
What I mean is if a application needs a database transaction you can't run another transaction/savepoint for your test.

the only way, which would work is to have something like if environment.mode = Mode.Test which won't do a commit; and the session with a running transaction would be overwritten by your test.

**Comment by chadjaros** [comment]
@schmitch It's possible to write your data access layer in such a way that the sessions are injected, the transactions managed from a higher layer. You would test the data access methods by providing that transacted session from the test, rolling it back at the end rather than committing. The transaction layer would then be tested via mocks.

I'm not certain how easily that can be achieved with Slick.

**Comment by chadjaros** [comment]
Arrgh. Mocking DBIOActions such that they work with for-comprehension (and subsequently anything with transactions) is turning out to be more trouble than it's worth.

I'm starting to think that Slick is more trouble than it's worth when advanced database features are necessary.

**Comment by asazernik** [comment]
@chadjaros What exactly are you referring to when you say "Mocking DBIOActions"?

I've got an idiom I use for non-testing code where I upgrade Trys and Futures to DBIOs:

```
  implicit def tryToFuture[A](t: Try[A]): Future[A] = t match {
    case Success(s) => Future.successful(s)
    case Failure(e) => Future.failed(e)
  }

  implicit def toDBIO[A](f: Future[A]): DBIOAction[A, NoStream, Effect] =
    DBIO.from(f)

  implicit def toDBIO[A](t: Try[A]): DBIOAction[A, NoStream, Effect] =
    toDBIO(tryToFuture(t))
```

Unfortunately implicit conversions don't work neatly with for comprehensions, since both Future and Try have their own flatMap implementations; I'm not sure what the exact implications are of nesting DBIOAction#flatMap inside Try#flatMap, but I doubt it would be good. Maybe a better idiom would be like this:

```scala
def  InXact[T](r : => T): DBIOAction[T, NoStream, Effect] = toDBIO(Try(r))
```

So a for comprehension could look like:

```scala
var baseAction = for {
  a <- SomeTable.filter(somePredicate).result
  b <- InXact {
    normalCodeThatCanThrowException(a)
    andMoreCode
  }
  c <- SomeTable += b
} yield (c == expectedValue)

wrapWithXactRollback(baseAction) // left as an exercise for the reader
```

**Comment by chadjaros** [comment]
I'm writing tests using JUnit and Mockito.

In my target method I had a block of code like:
    
    var combinedAction = for {
      _ <- myDao1.statement1(args)
      _ <- myDao2.statement2(args)
      _ <- myDao3.statement3(args)
    } yield ()

    Await.result(database.run(combinedAction), Duration(10, TimeUnit.SECONDS)

which lived in a piece of code. The myDao statement methods return `DBIOAction`, and in my unit tests myDao1-3 are mocked objects. Therefore, in order to unit test this block, I had the task of figuring out how to create a mock DBIOAction such that the for-comprehension would exercise all three method calls, like it would in a real execution of the block.

I was able to create a mock DBIOAction using Mockito, but I was unable to figure out a .flatmap implementation (including an implicit execution context) which would cause each of the three statements in the block to execute within a unit test. 

I gave up and changed the code to:

    Await.result(database.run(myDao1.statement1(args)), Duration(10, TimeUnit.SECONDS)
    Await.result(database.run(myDao2.statement2(args)), Duration(10, TimeUnit.SECONDS)
    Await.result(database.run(myDao3.statement3(args)), Duration(10, TimeUnit.SECONDS)

which behaves similarly and is much simpler to unit test.

This also means that any transactional blocks would be similarly difficult to unit test.

**Comment by asazernik** [comment]
Hmmm. Poking through the internals, I think it might be more productive to hack up the QueryExecutor in a given DatabaseConfig rather than the DBIOAction implementation. But "more" is a relative term - I think this would still be a pain.

**Comment by skozlov** [comment]
> As this seems to be a common migration problem, I support the integration of this feature. Probably it should be documented that this is not the recommended way to use Slick.

To me, it is not a migration problem. I don't want to make a "Hello, world" demonstrating how fast and cool Slick is. I want the DB access library to be a team player in my application. IMHO, it must be totally hidden inside the persistence layer, and the service layer should not be aware of it at all. But what if I need managing transactions at the service layer while keeping my repositories returning my entities, not `DBIOAction` or the other bowels? I can do it with Slick 2. Is it possible with Slick 3?

**Comment by szeiger** [comment]
This would have to be taken into account for #1564. I think that for an asynchronous API it is conceptionally impossible. If you do synchronous calls, you can keep state in the current thread, but if everything is asynchronous, that is no longer possible. You need to manage sequencing of actions explicitly (which is just what DBIOAction does).
